workers/unity/Packages/io.improbable.worker.sdk/ParameterConversion.cs

```diff
@@ -0,0 +1,140 @@
+using System;
+using System.Collections.Generic;
+using System.Runtime.ConstrainedExecution;
+using System.Runtime.InteropServices;
+
+namespace Improbable.Worker.CInterop.Internal
+{
+    // Merge with /worker_sdk/csharp_cinterop/sdk/worker/internal/ParameterConversion.cs
+    // when changes upstreamed to Worker SDK
+    internal static class ParameterConversion
+    {
+        internal class WrappedGcHandle : CriticalFinalizerObject, IDisposable
+        {
+            private GCHandle handle;
+
+            public WrappedGcHandle(object obj)
+            {
+                handle = GCHandle.Alloc(obj);
+            }
+
+            ~WrappedGcHandle()
+            {
+                if (handle.IsAllocated)
+                {
+                    handle.Free();
+                }
+            }
+
+            public void Dispose()
+            {
+                if (handle.IsAllocated)
+                {
+                    handle.Free();
+                }
+
+                GC.SuppressFinalize(this);
+            }
+
+            public IntPtr Get()
+            {
+                return GCHandle.ToIntPtr(handle);
+            }
+        }
+
+        public static unsafe CEventTrace.SpanId ConvertSpanId(SpanId spanId)
+        {
+            var internalSpanId = new CEventTrace.SpanId();
+            ApiInterop.Memcpy(internalSpanId.Data, spanId.Data, (UIntPtr) 16);
+            return internalSpanId;
+        }
+
+        public static unsafe CEventTrace.Event ConvertEvent(Event eventToConvert)
+        {
+            var internalEvent = new CEventTrace.Event();
+            internalEvent.UnixTimestampMillis = eventToConvert.UnixTimestampMillis;
+            internalEvent.Id = ConvertSpanId(eventToConvert.Id);
+            internalEvent.Data = eventToConvert.Data?.eventData.AddrOfPinnedObject() ?? IntPtr.Zero;
+
+            fixed (byte* eventType = ApiInterop.ToUtf8Cstr(eventToConvert.Type))
+            fixed (byte* eventMessage = ApiInterop.ToUtf8Cstr(eventToConvert.Message))
+            {
+                internalEvent.Type = eventType;
+                internalEvent.Message = eventMessage;
+            }
```

Comment:
This isn't safe as once we leave the `fixed` block, the pointers `eventType` and `eventMessage` may or may not be valid anymore. 

We will need to pin these using the handles and return the handles to the caller. 

Alternatively, have a look at `ConvertPlayerIdentityTokenRequest` or `ConvertLoginTokensRequest` in the Worker SDK for an different pattern where the caller provides a closure which will use the `internalEvent` while these pointers are still valid.

Reply:
I've had a go at using the same pattern as the `ConvertPlayerIdentityTokenRequest`